### PR TITLE
Fix some issues found by clang

### PR DESCRIPTION
**Pull request by avikivity**
This series fixes issues found by clang. Most are real issues that gcc just doesn't find, a few are due to clang lagging behind on some C++ updates. See individual explanations in patches.

The series is not sufficient to build with clang; it just addresses the simple problems. Two larger problems remain: clang isn't able to compile std::ranges (not clear yet whether this is a libstdc++ problem or a clang problem) and clang can't capture structured binding variables (due to lagging behind on the standard).

The motivation for building with clang is gaining access to a working implementation of coroutines and modules.

This series compiles with gcc and the unit tests pass.

### CODE REVIEW

File: test/boost/cql_query_test.cc [hunk 1]
```diff
@@ -3189,13 +3189,21 @@ SEASTAR_TEST_CASE(test_static_multi_cell_static_lists_with_ckey) {
             auto msg = e.execute_cql("SELECT slist, v FROM t WHERE p = 1 AND c = 1;").get0();
             auto slist_type = list_type_impl::get_instance(int32_type, true);
             assert_that(msg).is_rows().with_row({
-                { slist_type->decompose(make_list_value(slist_type, list_type_impl::native_type({3}))) },
+                { slist_type->decompose(make_list_value(slist_type, list_type_impl::native_type({data_value(3)}))) },
```
Comment (tgrabiec) [inline test/boost/cql_query_test.cc:3192]: Looks like an unrelated change

Comment (avikivity) [reply test/boost/cql_query_test.cc:3192]: yes, it's necessary, but should have been in its own patch. I'll split it off.

**Comment by avikivity** [comment]
Updated: new patch bf8c8d292a8199ab43201991d599ac3c400d84d6 split off from its host patch. No changes in diff.

**Review by tgrabiec — APPROVED** [review]

**Comment by avikivity** [comment]
@tgrabiec I see it's approved but not merged, what else is needed?

**Comment by psarna** [comment]
I'll merge

**Review by psarna — APPROVED** [review]

**Comment by psarna** [comment]
I'll temporarily dequeue due to this build failure:
```cpp
09:13:05  g++ -MD -MT build/release/table.o -MF build/release/table.o.d -I/jenkins/workspace/scylla-master/next/scylla/seastar/include -I/jenkins/workspace/scylla-master/next/scylla/build/release/seastar/gen/include -std=gnu++20 -U_FORTIFY_SOURCE -Wno-maybe-uninitialized -DSEASTAR_SSTRING -march=westmere -Werror=unused-result -DSEASTAR_API_LEVEL=4 -DSEASTAR_HAVE_DPDK -DFMT_LOCALE -DFMT_SHARED -I/usr/include/p11-kit-1 -O3 -ffunction-sections -fdata-sections -Wl,--gc-sections -Wstack-usage=13312 --param inline-unit-growth=300  -g -gz -iquote. -iquote build/release/gen -Wno-error=stack-usage=-ffile-prefix-map=/jenkins/workspace/scylla-master/next/scylla=.  -march=westmere -DBOOST_TEST_DYN_LINK  -I abseil -fvisibility=hidden -Wall -Werror -Wno-mismatched-tags -Wno-maybe-uninitialized -Wno-tautological-compare -Wno-missing-braces -Wno-misleading-indentation -Wno-overflow -Wno-noexcept-type -Wno-nonnull-compare -Wno-error=cpp -Wno-ignored-attributes -Wno-overloaded-virtual -Wno-stringop-overflow -Wno-error=deprecated-declarations -DXXH_PRIVATE_API -DSEASTAR_TESTING_MAIN -DHAVE_LZ4_COMPRESS_DEFAULT  -c -o build/release/table.o table.cc
09:13:05  In file included from table.cc:45:
09:13:05  utils/error_injection.hh:310:7: error: specialization of 'utils::error_injection<false>' after instantiation
09:13:05    310 | class error_injection<false> {
09:13:05        |       ^~~~~~~~~~~~~~~~~~~~~~
09:13:05  utils/error_injection.hh:310:7: error: redefinition of 'class utils::error_injection<false>'
09:13:05  In file included from table.cc:45:
09:13:05  utils/error_injection.hh:122:7: note: previous definition of 'class utils::error_injection<false>'
09:13:05    122 | class error_injection {
```

**Comment by avikivity** [comment]
Updated: 5cb3c5cf36c8136a337bfc9901222ea0fc6709bc rewritten to avoid build failure in release mode. It's sufficiently different that it merits re-review of that patch.

Tested build and unit tests in dev and release modes.

File: alternator/base64.cc [hunk 2]
```diff
@@ -32,11 +32,11 @@
 // and the character used in base64 encoding to represent it.
 static class base64_chars {
 public:
-    static constexpr const char* to =
+    static constexpr const char to[] =
             "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
     int8_t from[255];
     base64_chars() {
-        static_assert(strlen(to) == 64);
+        static_assert(sizeof(to) == 64 + 1);
```
Comment (nyh) [inline alternator/base64.cc:39]: Looks good. I don't understand your statement that strlen() can't be constexpr - the compilers are allowing more and more stuff to run in constexpr functions, and there is no reason not to have a constexpr version of strlen() that will work on constexpr strings. But I agree, if it's not implemented in clang, let's not use it, and you saw there is an easy workaround.

File: alternator/streams.cc [hunk 3]
```diff
@@ -77,11 +77,7 @@ struct from_string_helper {
     static ValueType& Set(ValueType& v, T data, typename ValueType::AllocatorType& a) {
         std::ostringstream ss;
         ss << data;
-#if __cplusplus >= 202002L
```
Comment (nyh) [inline alternator/streams.cc:80]: @elcallio can you please review this change to your code? Was the usage of view() good for performance?
ostringstream::view() was added in C++20, so probably the #if is somehow wrong, or clang implemented the standard wrong. I wonder if we could have some other different ifdef that will allow us to use the new view() when it's available.

File: cql3/statements/select_statement.cc [hunk 4]
```diff
@@ -551,7 +551,7 @@ indexed_table_select_statement::do_execute_base_query(
             concurrency *= 2;
             return proxy.query(_schema, command, std::move(prange), options.get_consistency(), {timeout, state.get_permit(), state.get_client_state(), state.get_trace_state()})
             .then([&ranges_to_vnodes, &merger] (service::storage_proxy::coordinator_query_result qr) {
-                bool is_short_read = qr.query_result->is_short_read();
+                auto is_short_read = qr.query_result->is_short_read();
```
Comment (nyh) [inline cql3/statements/select_statement.cc:554]: Why not add a regular, implicit conversion from bool_class to bool?
We of course won't have the opposite direction (the conversion of bool to bool_class is explicit, that's the whole point of bool_class), but why not allow implicit conversion from bool_class to bool?

File: sstables/size_tiered_backlog_tracker.hh [hunk 5]
```diff
@@ -89,7 +89,7 @@ class size_tiered_backlog_tracker final : public compaction_backlog_tracker::imp
     inflight_component compacted_backlog(const compaction_backlog_tracker::ongoing_compactions& ongoing_compactions) const;
 
     double log4(double x) const {
-        static constexpr double inv_log_4 = 1.0f / std::log(4);
+        double inv_log_4 = 1.0f / std::log(4);
```
Comment (nyh) [inline sstables/size_tiered_backlog_tracker.hh:92]: I think this change is ok, but it makes me feel uneasy:
I think if the compiler knows how to calculate / optimize this expression during compile time, that's the quintessential "constexpr", and actually writing constexpr was a nice way to emphasize that you insist that this optimization must be done, so it's sad clang doesn't allow us to use it. Additionally, the fact you removed the "static" means you count even more on this optimization being done (because if it's not, better do the actual calculation only once! that's what the "static" does). I hope you are right that it is. If we had a reason to suspect it isn't, keeping the "static" would have made sense.

File: test/boost/cql_query_test.cc [hunk 6]
```diff
@@ -3196,6 +3196,14 @@ SEASTAR_TEST_CASE(test_static_multi_cell_static_lists_with_ckey) {
     });
 }
 
+// std::pow() is not constexpr
+static constexpr int constexpr_int_pow(int x, int y) {
```
Comment (nyh) [inline test/boost/cql_query_test.cc:3200]: Should be "unsigned int y" - your implementation is (for obvious reasons) incorrect for negative y.

**Review by nyh — CHANGES_REQUESTED** [review]
Most of the patches look good. I only left a few comments about things which I thought might be a problem.

File: alternator/streams.cc (hunk 3 above)
Comment (elcallio) [reply alternator/streams.cc:80]: Docs only says c++20. The gcc headers use the above version guard as to whether declare it or not. I could not find any official version guard for the feature. The usage was only an optimization, yes (and to ease the pain I feel every day that not more parts of the standard lib uses views).

File: cql3/statements/select_statement.cc (hunk 4 above)
Comment (avikivity) [reply cql3/statements/select_statement.cc:554]: I don't know. Maybe it's worthwhile. I think implicit conversions are dangerous and it's better not to allow them. Maybe an implicit conversion to bool will also convert to other integrals.

File: sstables/size_tiered_backlog_tracker.hh (hunk 5 above)
Comment (avikivity) [reply sstables/size_tiered_backlog_tracker.hh:92]: constexpr is not about optimization. The compiler is free to emit runtime code to generate it. constexpr is about compile-time values, but it only forces compile-time evaluation if the value is used in a compile-time context.

File: test/boost/cql_query_test.cc (hunk 6 above)
Comment (avikivity) [reply test/boost/cql_query_test.cc:3200]: ok

File: alternator/base64.cc (hunk 2 above)
Comment (psarna) [reply alternator/base64.cc:39]: I think that in case of g++ it's not even that `strlen` has a constexpr variant, but rather the whole call is optimized out and replaced with a compile-time known constant.

**Comment by avikivity** [comment]
Updated: 70ea785cc7d9bc4258930051a85ddd61ca06fc31 now uses unsigned for the second parameter of constexpr pow.

File: alternator/base64.cc (hunk 2 above)
Comment (avikivity) [reply alternator/base64.cc:39]: Yes. g++ allowing it is a bug.

File: sstables/size_tiered_backlog_tracker.hh (hunk 5 above)
Comment (nyh) [reply sstables/size_tiered_backlog_tracker.hh:92]: I think it's clear that the intention of the author was to somehow guarantee that this 1/log(4) will not be calculated again and again on every call, and hence the use of "static" to make sure it's only calculated once, and "constexpr" as "fail this if it can't calculate at compile time". Your fix is ok - but it means that we just trust the optimizer to optimize this in any case - without all the tricks that the author tried to use. As optimizers are getting better (and slower), I guess we can just trust this...

File: cql3/statements/select_statement.cc (hunk 4 above)
Comment (nyh) [reply cql3/statements/select_statement.cc:554]: Well, somehow (?) we did have an implicit conversion to bool in gcc, and nobody complained, and your patch doesn't get rid of it. But never mind, your patch is fine too. We can rethink this later. I'll resolve this conversation.

File: alternator/streams.cc (hunk 3 above)
Comment (nyh) [reply alternator/streams.cc:80]: So my only remaining concern with this series is this. @elcallio do you think view() vs. str() has any performance implications? I think both of them may need to copy the string anyway (to make it contiguous) so it doesn't really matter which one we use?

Comment (elcallio) [reply alternator/streams.cc:80]: View does not need to copy anything. It just needs the original string (in the ostringstream) to live across its usage (which this does) - in fact since view is just a view, it prohibits actual data ownership. And method being const it pretty much implies no scattered buffers in iostreams. 

It is a minor perf diff. We will add a N byte copy, and probably extra malloc (since most objects serialized through here are > 32 chars), but it is hardly a deal breaker. Just a nice feature to have.

Comment (nyh) [reply alternator/streams.cc:80]: I'm not sure that the original ostringstream is actually guaranteed to contain a **contiguous** buffer, and if it doesn't, view() needs to copy as well. But it's indeed possible that it doesn't.
So, @elcallio if you're comfortable with Avi's patch which removes this C++20 optimization, I can commit this series.
If you're answer is no - you're not comfortable with it - I wonder what other trick (constraints? SFINAE? some other ifdef?) we could use to make this C++20 method work in gcc but not in clang which for some unknown (to me) reason doesn't implement it.

Comment (elcallio) [reply alternator/streams.cc:80]: I'm fine with it.

**Review by nyh — APPROVED** [review]

**Comment by nyh** [comment]
Committed to next: 73cb9e3f618013c61786b18308394a23e3e55974

**Comment by avikivity** [comment]
@nyh if you use `scripts/pull_pr.sh` it will note the queuing, and auto-close the pull request after it is merged.

**Comment by nyh** [comment]
> @nyh if you use `scripts/pull_pr.sh` it will note the queuing, and auto-close the pull request after it is merged.

Thanks. I never noticed this script before - I do something similar semi-manually. I'll check the script next time (I'm surprised it actually works,  it has some weird stuff like setting PR_REPO but never using it, and creating local branches that aren't cleaned up... but I'll test it).
Note that the two adventages you mentioned - noting the queuing and auto-close - are automatic Github features and have nothing to do with this script. And both of these do not happen at all when you commit a single-patch PR, which is why in any case I tend to do that manually.

**Comment by avikivity** [comment]
> Note that the two adventages you mentioned - noting the queuing and auto-close - are automatic Github features and have nothing to do with this script. And both of these do not happen at all when you commit a single-patch PR, which is why in any case I tend to do that manually.

Please check the script more carefully. It auto-closes single-patch pull requests too.

**Comment by nyh** [comment]
> Please check the script more carefully. It auto-closes single-patch pull requests too.

Ok, I didn't see that. I'm not sure I like the price for this (A "closes" of the PR, as if it was an actual issue) but what the heck. In any case, this wasn't a single-patch pull request. I'll try the script next time.

**Comment by avikivity** [comment]
> > Please check the script more carefully. It auto-closes single-patch pull requests too.
> 
> Ok, I didn't see that. I'm not sure I like the price for this (A "closes" of the PR, as if it was an actual issue) but what the heck. In any case, this wasn't a single-patch pull request. I'll try the script next time.

I don't like it either, but I prefer that to having to garbage-collect merged-but-unclosed pull requests.
